File: quick_sort/QuickSort.cs
/*
Time Complexity: Best: O(n log(n)), Average: O(n log(n)), Worst: O(n^2)
Space Complexity: O(log(n))
*/

using System;

public class QuickSort
{

    /// <summary>
    /// Chooses a pivot based on the median of three values in the array (start, mid, end - 1).
    /// Pivot is placed at the beginning of the array.
    /// </summary>
    /// <param name="a">The array</param>
    /// <param name="start">The start pivoting position in the array</param>
    /// <param name="end">The end pivoting position in the array</param>
    public static void ChoosePivot(int[] a, int start, int end)
    {
        int mid = start + (end - start) / 2;
        int median = Math.Max(Math.Min(a[start], a[mid]), Math.Min(Math.Max(a[start], a[mid]), a[end - 1]));
        int pivotPos = Array.IndexOf(a, median);
        Swap(a, pivotPos, start);
    }

    /// <summary>
    /// Partitions an array into two parts based on a pivot.
    /// All values smaller than the pivot will be placed before the pivot in the array.
    /// All values grater than the pivot will be placed after the pivot in the array
    /// </summary>
    /// <param name="a">The array</param>
    /// <param name="start">The start pivoting position in the array</param>
    /// <param name="end">The end pivoting position in the array</param>
    /// <returns>The new position of the pivot in the array</returns>
    public static int Partition(int[] a, int start, int end)
    {
        int pivotPos = start;       //pivot position (index in the array)
        int pivotVal = a[pivotPos]; //the value of the pivot
        int bigStart = start + 1;   //start index of all values greater than the pivot

        for (int curr = start + 1; curr < end; curr++)
        {
            if (a[curr] < pivotVal)
            {
                Swap(a, curr, bigStart);
                bigStart++;
            }
        }

        Swap(a, pivotPos, bigStart - 1);
        pivotPos = bigStart - 1;
        return pivotPos;
    }

    /// <summary>
    /// Sorts an array using the Quick Sort algorithm
    /// </summary>
    /// <param name="a">The array to sort</param>
    /// <param name="start">The start sorting position in the array</param>
    /// <param name="end">The end sorting position in the array</param>
    public static void DoQuickSort(int[] a, int start, int end)
    {
        if(end - start == 2)
        {
            if(a[start] > a[end - 1])
                Swap(a, start, end - 1);
        }
        else if(end - start > 1)
        {
            ChoosePivot(a, start, end);
            int pivotPos = Partition(a, start, end);
            DoQuickSort(a, start, pivotPos);
            DoQuickSort(a, pivotPos + 1, end);
        }
    }

    /// <summary>
    /// Quick Sort helper method
    /// </summary>
    /// <param name="a">The array to sort</param>
    public static void DoQuickSort(int[] a)
    {
        if (a != null && a.Length > 1)
            DoQuickSort(a, 0, a.Length);
    }

    /// <summary>
    /// Swap two index positions in an array
    /// </summary>
    /// <param name="a">The array</param>
    /// <param name="x">The first index</param>
    /// <param name="mid">The second index</param>
    public static void Swap(int[] a, int x, int y)
    {
        int temp = a[x];
        a[x] = a[y];
        a[y] = temp;
    }

    public static void Main()
    {
        int[] arr = new int[] {2, 4, 2, 6, 7, -1};
        DoQuickSort(arr);
        foreach(int element in arr)
        {
            Console.Write(element + " ");
        }
        Console.WriteLine("");
    }
}

File: quick_sort/quick_sort.py
def median(a, i, j, k):
    """
    Return median of 3 integers from array a.
    :param a: Iterable of elements
    :param i: start element index
    :param j: end element index
    :param k: middle element index
    :return: return median of values at indices i, j and k.
    """
    ai, aj, ak = a[i], a[j], a[k]
    med_val = ai + aj + ak - max(ai, aj, ak) - min(ai, aj, ak)
    if ai == med_val:
        return i
    elif aj == med_val:
        return j
    return k


def partition(array, l, r):
    """
    Perform Partition Operation on array.
    Time Complexity: Theta(nLogn)
    Auxiliary Space: O(n)
    :param array: Iterable of elements
    :param l: pivot value for array
    :param r: right limit of array
    :return: return q value for function, used in partitioning of array.
    """
    i = l - 1
    pivot_index = median(array, l, r, (l+r) // 2)
    array[pivot_index], array[r] = array[r], array[pivot_index]
    pivot = array[r]
    for j in range(l, r):
        if array[j] <= pivot:
            i += 1
            array[i], array[j] = array[j], array[i]
    i += 1
    array[r], array[i] = array[i], array[r]
    return i


def quick_sort(array, left, right):
    """
    Perform sort using partition function.
    Time Complexity : O(nlog(n)).
    Space Complexity : O(n).
    :param array: Iterable of elements
    :param left: used as left limit of quick sort
    :param right: right limit for quick sort
    :return: no returns, sorts array
    """
    if left < right:
        q = partition(array, left, right)
        quick_sort(array, left, q - 1)
        quick_sort(array, q + 1, right)


def main():
    a = [1, 2, 1, 2, 3, 1, 2, 2, 1]
    quick_sort(a, 0, len(a) - 1)
    print(a)


if __name__ == '__main__':
    main()
